### PR TITLE
Specified more GUI conventions

### LINKED ISSUE #1309: Look into defining conventions for Qt ui files and components that interact with them
### Description of the task

Quote from #1303 

> Maybe we should have a naming convention/guide for elements that interact with code, e.g. radioButton_ChargeMode

> Mmm good idea. I'll make an issue to look into this. We'll likely also want to experiment with the link between ui defined stuff and code, as well as nesting ui files.

### Acceptance criteria

- [ ] TODO

### LINKED ISSUE #1567: Add new GUI conventions to documentation
### Description of the task

* only Top-level components suffixed with `GUI`
* If we need custom functionality from anything, make a custom generic_widget to implement behavior
* `.ui` files should only be used for setting layout and organizing widgets (including custom ones). Don't set signals/slots here
* GUI classes should just aggregate top-level widgets (usually generated by the `.ui` file) and connect any callbacks / data. They shouldn't define any new widgets / features themselves, they are just top-level containers.

### Acceptance criteria

- [ ] Item A
- [ ] Item B
- [ ] Item C

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Specified a few more things for GUI conventions.

I moved the GUI stuff to its own section now that it's more generic beyond the Visualizer (the only part of the diff that actually changed is stuff under the GUI heading. The SimulatedTest stuff didn't change)

### Testing Done

N/A

### Resolved Issues

resolves #1309 
resolves #1567 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: docs/software-architecture-and-design.md [hunk 1]
```diff
@@ -446,7 +448,78 @@ Although we want to display information about the [AI](#ai) in the [Visualizer](
...
+
+Ultimately, we want a test suite that validates our [Plays](#plays) are generally doing the right thing. We might not care exactly where a robot receives the ball during a passing play, as long as the pass was successful overall. The solution to this problem is to use simulation to allow us to deterministically run our entire AI pipeline and validate behaviour.
+
+The primary design goals of this test system are:
+1. **Determinism:** We need tests to pass or fail consistently
+2. **Test "ideal" behaviour:** We want to test the logic in a "perfect world", where we don't care about all the exact limitations of our system in the real world with real physics. Eg. we don't care about modelling robot wheels slipping on the ground as we accelerate.
+3. **Ease of use:** It should be as easy and intuitive as possible to write tests, and understand what they are testing.
```
Comment (kelvinkoon) [inline docs/software-architecture-and-design.md]: "to write tests, and understand" -> "to write tests, and **to** understand"

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Pretty clear on first glance. Will take a more detailed look later.

File: docs/software-architecture-and-design.md (hunk 1 above)
Comment (MathewMacDougall) [reply docs/software-architecture-and-design.md]: fixed

File: docs/software-architecture-and-design.md [hunk 2]
```diff
@@ -446,7 +448,78 @@ Although we want to display information about the [AI](#ai) in the [Visualizer](
...
+
+Names should follow our [reglar naming conventions for variables](code-style-guide.md#names-and-variables). Furthermore, they should be named in the form `<purpose>_<widget>`. The purpose is essentially the "normal" variable name that should describe what the variable is. The widget component should be the name of the widget.
+
+For example, a `QLabel` for team colour should be named `team_colour_label`. Similarly, the button that starts the AI should be named `start_ai_button`.
+
+![Good Qt Widget Naming](images/qt_widget_naming_example.png)
+
+### Classes
+Only the top-level class for a given GUI should be prefixed with `GUI`. These top-level classes should just aggregate top-level widgets (typically generated from a `.ui` file) and connect callbacks. They should not define additional widgets or features themselves. For example, `FullSystemGUI` connects incoming buffers of data to the `main_widget` generated by a `.ui` file.
```
Comment (jonathanlew) [inline docs/software-architecture-and-design.md]:
```suggestion
Only the top-level class for a given GUI should be suffixed with `GUI`. These top-level classes should just aggregate top-level widgets (typically generated from a `.ui` file) and connect callbacks. They should not define additional widgets or features themselves. For example, `FullSystemGUI` connects incoming buffers of data to the `main_widget` generated by a `.ui` file.
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
The codebase isn't consistent with the conventions

**Comment by MathewMacDougall** [comment]
> The codebase isn't consistent with the conventions

Figured we should lock down convention first then update codebase separately?

**Review by LiCody — APPROVED** [review]
LGTM 👍

**Comment by MathewMacDougall** [comment]
I opened #1726 to update the code to match. @LiCody My PR just updated the `Simulator` and `FullSystemGUI`. Can you do a separate PR for the `RobotDiagnostics` since you're more familiar with it / what things should be named?

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Comment by MathewMacDougall** [comment]
@kelvinkoon need your review

**Review by kelvinkoon — APPROVED** [review]

**Comment by LiCody** [comment]
> I opened #1726 to update the code to match. @LiCody My PR just updated the `Simulator` and `FullSystemGUI`. Can you do a separate PR for the `RobotDiagnostics` since you're more familiar with it / what things should be named?

year sure, opened #1730

**Pull request merged**